src/main/java/uk/ac/ebi/embl/converter/gff3/reader/GFF3FileReader.java

```diff
@@ -0,0 +1,176 @@
+
+                if (parentId.isPresent()) {
+                    if (featureIdx.containsKey(parentId.get())) {
+                        featureIdx.get(parentId.get()).addChild(feature);
+                    }
+                    // TODO: Validation check, no parent found
+                } else if (attributesMap.containsKey("gene")) {
+                    String gene = attributesMap.get("gene");
+                    geneMap.putIfAbsent(gene, new ArrayList<>());
+                    List<GFF3Feature> geneFeatures = geneMap.get(gene);
+                    geneFeatures.add(feature);
+                } else {
+                    nonGeneFeatures.add(feature);
+                }
+                // TODO: Validation check accession matches sequence region.
+
+                continue;
+            }
+            break;
+        }
+
+        if (directives.isEmpty() && geneMap.isEmpty() && nonGeneFeatures.isEmpty()) {
+            return null;
+        }
+        return new GFF3Annotation(new GFF3Directives(directives), geneMap, nonGeneFeatures);
+    }
+
+    private static Map<String, String> attributesFromString(String line) {
+        Map<String, String> attributes = new HashMap<>();
+        String[] parts = line.split(";");
+        for (String part : parts) {
+            String[] keyValue = part.split("=");
+            attributes.put(keyValue[0], keyValue[1]);
+        }
+        return attributes;
+    }
+
+    private List<GFF3Annotation> readAnnotationList() throws IOException, GFF3ValidationError, GFF3ValidationError {
+        List<GFF3Annotation> annotations = new ArrayList<>();
+        while (true) {
+            // readAnnotation
+            GFF3Annotation annotation = readAnnotation();
+            if (annotation == null) {
+                break;
+            } else {
+                annotations.add(annotation);
+            }
+        }
+        return annotations;
+    }
+
+    public GFF3Header readHeader() throws IOException, GFF3ValidationError, GFF3ValidationError {
+        while (true) {
+            String line = this.bufferedReader.readLine();
+            if (line == null) {
+                throw new GFF3ValidationError("GFF3 header not found");
+            }
+            if (line.isBlank()) {
+                continue;
+            }
+
+            Matcher m = DIRECTIVE_VERSION.matcher(line);
+            if (m.matches()) {
+                String version = m.group("version");
+                GFF3Header header = new GFF3Header(version);
+                return header;
+            } else {
+                throw new GFF3ValidationError("Invalid GFF3 header");
+            }
+        }
+    }
+
+    public GFF3File read() throws IOException, GFF3ValidationError {
+        GFF3Header header = readHeader();
+        List<GFF3Annotation> annotations = this.readAnnotationList();
+        GFF3File gff3File = new GFF3File(header, annotations);
+        return gff3File;
```

Comment:
This code can be efficient. With this implementation, we end up reading the GFF3 file 2 times in the lines below: 
```
GFF3Header header = readHeader();
 List<GFF3Annotation> annotations = this.readAnnotationList();
```
Gff3 file must be read only once.